Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,99 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+fileprivate enum ClientState: Hashable {
+    case ready
+    case waitingForAuthenticationMethod(ClientGreeting)
+    case waitingForClientRequest
+    case waitingForServerResponse(ClientRequest)
+    case active
+}
+
+enum ClientAction: Hashable {
+    case none
+    case sendRequest
+    case proxyEstablished
+}
+
+public struct InvalidAuthenticationSelection: Error {
+    
+}
+
+public struct ConnectionFailed: Error, Hashable {
+    public var reply: Reply
+}
+
+struct ClientStateMachine {
+
+    private var state: ClientState
+    
+    var proxyEstablished: Bool {
+        switch self.state {
+        case .active:
+            return true
+        case .ready, .waitingForAuthenticationMethod, .waitingForClientRequest, .waitingForServerResponse:
+            return false
+        }
+    }
+    
+    init() {
+        self.state = .ready
+    }
+
+    mutating func sendClientGreeting(_ greeting: ClientGreeting) {
+        assert(self.state == .ready)
+        self.state = .waitingForAuthenticationMethod(greeting)
+    }
+    
+    mutating func sendClientRequest(_ request: ClientRequest) {
+        assert(self.state == .waitingForClientRequest)
+        self.state = .waitingForServerResponse(request)
+    }
+    
+    mutating func receiveBuffer(_ buffer: inout ByteBuffer) throws -> ClientAction? {
```

Comment:
The `ClientAction` has the option `.none`. Do we still want to have this method return an optional?

Reply:
I'm using `ClientAction.none` to mean "no action", whereas returning `nil`, means "not enough data", similar to B2MD (which I don't think is suitable to use here).

Reply:
Okay two things:
1. @Lukasa might have opinions on parsing in state machines. After his [great talk](https://www.youtube.com/watch?v=7cC3_jGwl_U) I think I'd prefer to have the parsing extra. (This probably makes the state machine easier to reason about as well)
2. If we keep the parsing in the SM, we could have an action `.waitForMoreData` that would be semantically what nil is as of now.

Reply:
In the case of SOCKS it is not possible to have a parser that is unaware of connection state: the messages are not self-framing and so we don't know how to parse them without knowing where we are in the connection state. That is a strong argument for having the state machine own the parser: it avoids duplicating that state.

Reply:
@Davidde94 what do you think about the `.waitForMoreData` action?

Reply:
Personally I think `nil` is a bit neater. I'll make the change for now and we can bike shed a little, @Lukasa can have the deciding vote!

Reply:
(change is pushed up)

Reply:
The implementation has changed quite a bit and I think we can call this conversation resolved. @fabianfett reopen if you're not happy!